bot/tools/changelog/main.go

```diff
@@ -0,0 +1,220 @@
+	}
+	entCLGen := &changelogGenerator{
+		isEnt: true,
+		dir:   entDir,
+	}
+	ossCL, err := ossCLGen.generateChangelog(branch, timeLastRelease, timeNow)
+	if err != nil {
+		log.Fatal(err)
+	}
+	entCL, err := entCLGen.generateChangelog(branch, timeLastEntRelease, timeLastEntMod)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	fmt.Println(ossCL)
+	if entCL != "" {
+		fmt.Println("Enterprise:")
+		fmt.Println(entCL)
+	}
+}
+
+func prereqCheck() error {
+	if err := git.IsAvailable(); err != nil {
+		return trace.Wrap(err)
+	}
+	if err := gh.IsAvailable(); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// getBranch will return branch if parsed otherwise will attempt to find it
+// Branch should be in the format "branch/v*"
+func getBranch(branch, dir string) (string, error) {
+	if branch != "" {
+		return branch, nil
+	}
+	// get ref
+	ref, err := git.RunCmd(dir, "symbolic-ref", "HEAD")
+	if err != nil {
+		return "", trace.Wrap(err, "not on a branch")
+	}
+
+	// remove prefix and ensure that branch is in expected format
+	branch, _ = strings.CutPrefix(ref, "refs/heads/")
+	if branch == ref {
+		return "", trace.BadParameter("not on a branch: %s", ref)
+	}
+
+	// if the branch is not in the branch/v* format then check it's root
+	if !strings.HasPrefix(branch, "branch/v") {
+		fbranch, err := getForkedBranch(dir)
+		if err != nil {
+			return "", trace.Wrap(err, "could not determine a root branch")
+		}
+		branch = fbranch
+	}
+
+	return branch, nil
+}
+
+// getForkedBranch will attempt to find a root branch for the current one that is in the format branch/v*
+func getForkedBranch(dir string) (string, error) {
+	forkPointRef, err := git.RunCmd(dir, "merge-base", "--fork-point", "HEAD")
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	fbranch, err := git.RunCmd(dir, "branch", "--list", "branch/v*", "--contains", forkPointRef, "--format", "%(refname:short)")
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	if fbranch == "" { // stdout is empty indicating the search failed
+		return "", trace.Errorf("could not find a valid root branch")
+	}
+	return fbranch, nil
+}
+
+func getLastVersion(baseTag, dir string) (string, error) {
+	if baseTag != "" {
+		return baseTag, nil
+	}
+
+	// get root dir of repo
+	topDir, err := git.RunCmd(dir, "rev-parse", "--show-toplevel")
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	lastVersion, err := makePrintVersion(topDir)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	return lastVersion, nil
+}
+
+// makePrintVersion will run 'make -s print-version'
+func makePrintVersion(dir string) (string, error) {
+	var stdout bytes.Buffer
+	var stderr bytes.Buffer
+
+	cmd := exec.Command("make", "-s", "print-version")
+	cmd.Dir = dir
+	cmd.Stderr = &stderr
+	cmd.Stdout = &stdout
+
+	err := cmd.Run()
+
+	if err != nil {
+		return strings.TrimSpace(stderr.String()), trace.Wrap(err, "can't get last released version")
+	}
+	out := strings.TrimSpace(stdout.String())
+
+	return "v" + out, nil
+}
```

Comment:
IMO it would be better to pass this in as a parameter rather than invoking `make`. This would allow the tool to be used in other projects without this target.

Reply:
So this is actually passed in as a flag (`--base-tag`). This is just the fallback if that flag isn't passed in. I tried to keep the functionality close to the current implementation. But yeah this would fail if the `--base-tag` flag isn't set and the Makefile doesn't exist or have that target making it not so portable.

For now to be clearer e.g. `--base-tag was not set, defaulted to invoking 'make version' but failed`. Should be less ambiguous and indicate a way to "fix" it.